jupyter-extension/src/plugins/mount/mountDrive.ts

```diff
@@ -1,61 +1,123 @@
   modelDBFactory?: ModelDB.IFactory | undefined;
 
   constructor(registry: DocumentRegistry) {
     this._registry = registry;
+    this._model = {page: 0, max_page: 0};
+    this._cache = {key: null, contents: DEFAULT_CONTENT_MODEL};
   }
 
   get name(): 'mount-browser' {
     return 'mount-browser';
   }
+
+  get model(): Paging.IModel {
+    return this._model;
+  }
+
   get fileChanged(): ISignal<this, Contents.IChangedArgs> {
     return this._fileChanged;
   }
 
+  // Signal emits `true` on expensive request to backend and `false` on response.
+  get loading(): ISignal<this, boolean> {
+    return this._loading;
+  }
+
   get serverSettings(): ServerConnection.ISettings {
     return this._serverSettings;
   }
   get isDisposed(): boolean {
     return this._isDisposed;
   }
 
+  async _get(
+    url: string,
+    options: PartialJSONObject,
+  ): Promise<Contents.IModel> {
+    url += URLExt.objectToQueryString(options);
+    return requestAPI<Contents.IModel>(url, 'GET');
+  }
+
+  // We might be able to pass pagination to the ContentsManager backend by including
+  // it in the URI specifier. This would require a custom ContentsManager which
+  // supports pagination.
   async get(
     localPath: string,
     options?: Contents.IFetchOptions,
   ): Promise<Contents.IModel> {
+    const url = URLExt.join('pfs', localPath);
+
+    // Make a no-content request to determine the type of content being requested.
+    // If the content type is not a directory, then we want to preserve our cache
+    //   and not paginate the results.
+    let shallowResponse;
     try {
-      const response = await requestAPI<Contents.IModel>(
-        URLExt.join('pfs', localPath),
-        'GET',
-      );
-      return response;
+      shallowResponse = await this._get(url, {content: '0'});
     } catch (e) {
       console.log('/pfs not found');
-      return {
-        name: '',
-        path: '',
-        last_modified: '2022-04-26T16:28:48.015858Z',
-        created: '2022-04-26T16:28:48.015858Z',
-        content: [],
-        format: 'json',
-        mimetype: '',
-        size: undefined,
-        writable: true,
-        type: 'directory',
-      };
+      return DEFAULT_CONTENT_MODEL;
+    }
+    const content = options?.content ? '1' : '0';
+    if (content === '0') {
+      return shallowResponse;
     }
+
+    if (shallowResponse.type !== 'directory') {
+      return await this._get(url, {...options, content});
+    }
+
+    // If we don't have contents cached, then we fetch them and cache the results.
+    if (localPath !== this._cache.key || !localPath) {
+      this._loading.emit(true);
+      const response = await this._get(url, {...options, content});
+      this._loading.emit(false);
+      this._model.page = 1;
+      this._model.max_page = Math.ceil(
+        response.content.length / MAX_NUM_CONTENTS_PAGE,
+      );
+      this._cache = {key: localPath, contents: response.content};
```

Comment:
If I understand correctly, we only replace the cache contents if `localPath` is a directory. This means if we click into a subdirectory, we replace the cache, so when we later back up on level, we'll have to re-fetch. Just wanted to mention this. I think I'm in favor of simple logic that just caches the current directory over a multi-index cache too.

Reply:
Exactly, the cache is only for the directory the file browser is currently displaying. When it cd's into another directory, the cache is updated. The reason why I chose this simple cache as opposed to a multi-index cache is to prevent memory leaks. Since the contents manager always sends back all the contents of the directory, all that data is unavoidably loaded into memory and so there's no harm in caching it to speed up pagination.